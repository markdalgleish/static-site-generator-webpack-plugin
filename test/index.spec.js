const path = require('path');
const { promisify } = require('es6-promisify');
const clean = require('rimraf');
const dirCompare = require('dir-compare');
const webpack = promisify(require('webpack'));
const rimrafAsync = promisify(require('rimraf'));
const getSubDirsSync = require('./utils/get-sub-dirs-sync');
const dirContentsToObject = require('./utils/dir-contents-to-object');
const directoryContains = require('./utils/directory-contains');

const successCases = getSubDirsSync(__dirname + '/success-cases');
const errorCases = getSubDirsSync(__dirname + '/error-cases');

jest.setTimeout(20000);

describe('Success cases', () => {
  getSubDirsSync(__dirname + '/success-cases').forEach(testCaseName => {
    describe(testCaseName, () => {
      const testCaseRoot = __dirname + '/success-cases/' + testCaseName;

      beforeAll(() => rimrafAsync(testCaseRoot + '/actual-output/'));

      it('should generate the expected files', () => {
        return webpack(require(path.join(testCaseRoot, 'webpack.config.js')))
          .then(() => dirContentsToObject(testCaseRoot + '/actual-output'))
          .then(files => {
            expect(files).toMatchSnapshot();
          });
      });
    });
  });
});

describe('Error cases', () => {
  errorCases.forEach(errorCase => {
    describe(errorCase, () => {
      beforeEach(() =>
        rimrafAsync(__dirname + '/error-cases/' + errorCase + '/actual-output')
      );

      it('generates the expected error', () => {
        const webpackConfig = require('./error-cases/' +
          errorCase +
          '/webpack.config.js');
        const expectedError = require('./error-cases/' +
          errorCase +
          '/expected-error.js');

        return webpack(webpackConfig).then(stats => {
          const actualError = stats.compilation.errors[0]
            .toString()
            .split('\n')[0];
          expect(actualError).toContain(expectedError);
        });
      });
    });
  });
});
