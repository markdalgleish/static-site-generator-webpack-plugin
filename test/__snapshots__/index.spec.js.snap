// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Success cases basic should generate the expected files 1`] = `
Object {
  "foo/bar/index.html": "<html>
  <body>
    <h1>/foo/bar</h1>
  </body>
</html>
",
  "foo/index.html": "<html>
  <body>
    <h1>/foo</h1>
  </body>
</html>
",
  "index.html": "<html>
  <body>
    <h1>/</h1>
  </body>
</html>
",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "stats.json": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases basic-promise should generate the expected files 1`] = `
Object {
  "foo/bar/index.html": "<html>
  <body>
    <h1>/foo/bar</h1>
  </body>
</html>
",
  "foo/index.html": "<html>
  <body>
    <h1>/foo</h1>
  </body>
</html>
",
  "index.html": "<html>
  <body>
    <h1>/</h1>
  </body>
</html>
",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "stats.json": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases basic-sync should generate the expected files 1`] = `
Object {
  "foo/bar/index.html": "<html>
  <body>
    <h1>/foo/bar</h1>
  </body>
</html>
",
  "foo/index.html": "<html>
  <body>
    <h1>/foo</h1>
  </body>
</html>
",
  "index.html": "<html>
  <body>
    <h1>/</h1>
  </body>
</html>
",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "stats.json": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases compression should generate the expected files 1`] = `
Object {
  "index.html": "<html>
  <body>
    <h1>/</h1>
  </body>
</html>
",
  "index.html.gz": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases crawl-ignore should generate the expected files 1`] = `
Object {
  "index.html": "<p>Don't crawl these:</p>
<ul>
  <li><a href=\\"http://webpack.js.org\\">Absolute Link</a></li>
  <li><a href=\\"//webpack.js.org\\">Protocol Relative Link</a></li>
  <li><a href=\\"mailto:hello@example.com\\">Email Link</a></li>
  <li><a href=\\"#hashtown\\">Hash Link</a></li>
</ul>
",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases crawl-relative should generate the expected files 1`] = `
Object {
  "foo/bar/baz/index.html": "<h1>/foo/bar/baz</h1>

<a href=\\"../../../\\">Crawl this</a>
",
  "foo/bar/index.html": "<h1>/foo/bar</h1>

<a href=\\"./bar/baz\\">Crawl this</a>
",
  "foo/index.html": "<h1>/foo</h1>

<a href=\\"./foo/bar\\">Crawl this</a>
",
  "index.html": "<h1>/</h1>

<a href=\\"foo\\">Crawl this</a>
",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases crawl-relative-iframes should generate the expected files 1`] = `
Object {
  "foo/bar/baz/index.html": "<h1>/foo/bar/baz</h1>

<iframe src=\\"javascript:void(0)\\" />
",
  "foo/bar/index.html": "<h1>/foo/bar</h1>

<iframe src=\\"./bar/baz\\" />
",
  "foo/index.html": "<h1>/foo</h1>

<iframe src=\\"./foo/bar\\" />
",
  "index.html": "<h1>/</h1>

<iframe src=\\"./foo\\" />
",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases crawl-root-relative should generate the expected files 1`] = `
Object {
  "foo/bar/index.html": "<h1>/foo/bar</h1>

<a href=\\"/\\">Crawl this</a>
",
  "foo/index.html": "<h1>/foo</h1>

<a href=\\"/foo/bar\\">Crawl this</a>
",
  "index.html": "<h1>/</h1>

<a href=\\"/foo\\">Crawl this</a>
",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases crawl-root-relative-iframes should generate the expected files 1`] = `
Object {
  "foo/bar/index.html": "<h1>/foo/bar</h1>

<iframe src=\\"javascript:void(0)\\" />
",
  "foo/index.html": "<h1>/foo</h1>

<iframe src=\\"/foo/bar\\" />
",
  "index.html": "<h1>/</h1>

<iframe src=\\"/foo\\" />
",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases custom-file-names should generate the expected files 1`] = `
Object {
  "custom.html": "<html>
  <body>
    <h1>/custom.html</h1>
  </body>
</html>
",
  "foo/bar/custom.html": "<html>
  <body>
    <h1>/foo/bar/custom.html</h1>
  </body>
</html>
",
  "foo/custom.html": "<html>
  <body>
    <h1>/foo/custom.html</h1>
  </body>
</html>
",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases default-entry-when-named should generate the expected files 1`] = `
Object {
  "foo/bar/index.html": "<html>
  <body>
    <h1>/foo/bar</h1>
  </body>
</html>
",
  "foo/index.html": "<html>
  <body>
    <h1>/foo</h1>
  </body>
</html>
",
  "index.html": "<html>
  <body>
    <h1>/</h1>
  </body>
</html>
",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "stats.json": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases entry-via-file-name should generate the expected files 1`] = `
Object {
  "foo/bar/index.html": "<html>
  <body>
    <h1>/foo/bar</h1>
  </body>
</html>
",
  "foo/index.html": "<html>
  <body>
    <h1>/foo</h1>
  </body>
</html>
",
  "index.html": "<html>
  <body>
    <h1>/</h1>
  </body>
</html>
",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "stats.json": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases entry-via-name should generate the expected files 1`] = `
Object {
  "foo/bar/index.html": "<html>
  <body>
    <h1>/foo/bar</h1>
  </body>
</html>
",
  "foo/index.html": "<html>
  <body>
    <h1>/foo</h1>
  </body>
</html>
",
  "index.html": "<html>
  <body>
    <h1>/</h1>
  </body>
</html>
",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "stats.json": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases es-modules should generate the expected files 1`] = `
Object {
  "foo/bar/index.html": "<html>
  <body>
    <h1>/foo/bar</h1>
  </body>
</html>
",
  "foo/index.html": "<html>
  <body>
    <h1>/foo</h1>
  </body>
</html>
",
  "index.html": "<html>
  <body>
    <h1>/</h1>
  </body>
</html>
",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "index.js.map": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases ignore-inherited-locals should generate the expected files 1`] = `
Object {
  "index.html": "<html>
  <body>
    <h1>\\"DONT_INCLUDE_ME\\" is in locals: false</h1>
  </body>
</html>
",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases legacy-args should generate the expected files 1`] = `
Object {
  "1.index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "foo/bar/index.html": "<html>
  <body>
    <h1>/foo/bar</h1>
  </body>
</html>
",
  "foo/index.html": "<html>
  <body>
    <h1>/foo</h1>
  </body>
</html>
",
  "index.html": "<html>
  <body>
    <h1>/</h1>
  </body>
</html>
",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases multi-render should generate the expected files 1`] = `
Object {
  "foo/bar/index.html": "<html>
  <body>
    <h1>/foo/bar</h1>
  </body>
</html>
",
  "foo/index.html": "<html>
  <body>
    <h1>/foo</h1>
  </body>
</html>
",
  "index.html": "<html>
  <body>
    <h1>/</h1>
  </body>
</html>
",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "stats.json": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases multi-render-custom-file-names should generate the expected files 1`] = `
Object {
  "custom.html": "<html>
  <body>
    <h1>/custom.html</h1>
  </body>
</html>
",
  "foo/bar/custom.html": "<html>
  <body>
    <h1>/foo/bar/custom.html</h1>
  </body>
</html>
",
  "foo/custom.html": "<html>
  <body>
    <h1>/foo/custom.html</h1>
  </body>
</html>
",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases no-options-provided should generate the expected files 1`] = `
Object {
  "index.html": "[\\"path\\",\\"assets\\",\\"webpackStats\\"]",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "stats.json": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases require-ensure should generate the expected files 1`] = `
Object {
  "1.index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "foo/bar/index.html": "<html>
  <body>
    <h1>/foo/bar</h1>
  </body>
</html>
",
  "foo/index.html": "<html>
  <body>
    <h1>/foo</h1>
  </body>
</html>
",
  "index.html": "<html>
  <body>
    <h1>/</h1>
  </body>
</html>
",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases single-path should generate the expected files 1`] = `
Object {
  "index.html": "<html>
  <body>
    <h1>/</h1>
  </body>
</html>
",
  "index.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`Success cases transform-locals should generate the expected files 1`] = `
Object {
  "bar.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "index.html": "<div>/foo/bar.js<br />/foo/main.js</div>",
  "main.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;
